src/rosdistro/manifest_provider/cache.py

```diff
@@ -130,13 +130,14 @@ def __call__(self, repo):
             logger.debug('Load package XMLs for repo "%s" from cache' % repo.name)
 
         # De-duplicate with the release package XMLs. This will cause the YAML writer
-        # to use references, saving a lot of space in the cache file.
-        for key in repo_cache or {}:
-            if key[0] != '_':
-                repo_cache[key][1] = sanitize_xml(repo_cache[key][1])
-                if key in self._distribution_cache.release_package_xmls:
-                    release_package_xml = self._distribution_cache.release_package_xmls[key]
-                    if repo_cache[key][1] == release_package_xml:
-                        repo_cache[key][1] = release_package_xml
+        # to use references for the common strings, saving a lot of space in the cache file.
+        if repo_cache:
+            for package_name, package_path, package_xml in repo_cache.items():
+                package_xml = sanitize_xml(package_xml)
+                release_package_xml = self._distribution_cache.release_package_xmls.get(package_name, None)
+                if package_xml == release_package_xml:
+                    # Re-add to the cache with a reference to the "release" instance of the string, so that
+                    # the YAML writer is able to notice that the same object appears multiple times.
+                    repo_cache.add(package_name, package_path, release_package_xml)
```

Comment:
The overall logic here is the same as previously, but it's been made a bit more explicit and non-magical.